public/app/features/alerting/unified/hooks/useProduceNewRuleGroup.tsx

```diff
@@ -0,0 +1,119 @@
+import { Action } from '@reduxjs/toolkit';
+import { useCallback, useState } from 'react';
+
+import { dispatch, getState } from 'app/store/store';
+import { RuleGroupIdentifier } from 'app/types/unified-alerting';
+import { RulerGrafanaRuleDTO, RulerRuleDTO, RulerRuleGroupDTO } from 'app/types/unified-alerting-dto';
+
+import { AlertGroupUpdated, alertRuleApi } from '../api/alertRuleApi';
+import { deleteRuleAction, pauseRuleAction, ruleGroupReducer } from '../reducers/ruler/ruleGroups';
+import { fetchRulesSourceBuildInfoAction, getDataSourceRulerConfig } from '../state/actions';
+
+type ProduceResult = RulerRuleGroupDTO | AlertGroupUpdated;
+
+// @TODO the manual state tracking here is not great, but I don't have a better idea that works /shrug
+function useProduceNewRuleGroup() {
+  const [fetchRuleGroup] = alertRuleApi.endpoints.getRuleGroupForNamespace.useLazyQuery();
+  const [updateRuleGroup] = alertRuleApi.endpoints.updateRuleGroupForNamespace.useMutation();
+  const [deleteRuleGroup] = alertRuleApi.endpoints.deleteRuleGroupFromNamespace.useMutation();
+
+  const [isLoading, setLoading] = useState<boolean>(false);
+  const [isUninitialized, setUninitialized] = useState<boolean>(true);
+  const [result, setResult] = useState<ProduceResult | undefined>();
+  const [error, setError] = useState<unknown | undefined>();
+
+  const isError = Boolean(error);
+  const isSuccess = !isUninitialized && !isLoading && !isError;
+
+  const requestState = {
+    isUninitialized,
+    isLoading,
+    isSuccess,
+    isError,
+    result,
+    error,
+  };
+
+  const produceNewRuleGroup = async (ruleGroup: RuleGroupIdentifier, action: Action) => {
+    const { dataSourceName, groupName, namespaceName } = ruleGroup;
+
+    // @TODO we should really not work with the redux state (getState) here
+    await dispatch(fetchRulesSourceBuildInfoAction({ rulesSourceName: dataSourceName }));
+    const rulerConfig = getDataSourceRulerConfig(getState, dataSourceName);
+
+    setUninitialized(false);
+    setLoading(true);
+
+    try {
+      const currentRuleGroup = await fetchRuleGroup({
```

Comment:
Likewise here, probably worth a comment to explain why we _have_ to fetch this fresh to ensure we're always going to use the full group definition from the API, before we try and persist any changes?

Reply:
I would rename this variable to something like `freshRuleGroup` to make it more explicit that this is a fresh group

Reply:
👍 solid feedback, I'll disambiguate the variable name and add some comments here